L1Trigger/Phase2L1ParticleFlow/interface/regionizer/tdr_regionizer_elements_ref.icc

```diff
@@ -1,325 +1,582 @@
+    timeOfNextObject_ = objClk;
+  }
 }
-//explicit for tracker to handle clocking
+
+template <typename T>
+void l1ct::tdr_regionizer::Buffer<T>::updateNextObjectTime(int currTime) {
+  if (data_.size() > 0) {
+    timeOfNextObject_ = std::max(front().clock(), static_cast<unsigned int>(currTime + 1));
+  } else {
+    timeOfNextObject_ = -1;
+  }
+}
+
+template <typename T>
+inline unsigned int l1ct::tdr_regionizer::Buffer<T>::nextObjClk(unsigned int ndup) {
+  unsigned int nextVal = std::max(clkindex360_, clkindex240_) / 3;
+
+  clkindex360_ += 2 * ndup;
+  clkindex240_ += 3;
+
+  return nextVal;
+}
+
+// explicit for tracks
 template <>
-inline void l1ct::tdr_regionizer::Pipe<l1ct::TkObjEmu>::addObj(l1ct::TkObjEmu obj,
-                                                               unsigned int phiindex,
-                                                               unsigned int etaindex,
-                                                               bool phioverlap,
-                                                               bool etaoverlap,
-                                                               int glbphi,
-                                                               int glbeta) {
-  data_.emplace_back(
-      PipeObject<l1ct::TkObjEmu>(obj, phiindex, etaindex, phioverlap, etaoverlap, glbphi, glbeta, clkindex_++));
-  if (clkindex_ % 3 == 2)
-    clkindex_++;  //this is for tracker, could I get this generically maybe?
-}
-
-template <typename T>
-int l1ct::tdr_regionizer::Pipe<T>::getClosedIndexForObject(unsigned int index) {
-  switch (getCount(index)) {
-    case 0:
-      return getPhi(index) * 2 + getEta(index);
-    case 1:
-      if (getPhiOverlap(index) && !getEtaOverlap(index))
-        return ((getPhi(index) + 1) % nphi_) * 2 + getEta(index);
-      else  //eta overlap, or 4-small-region overlap
-        return getPhi(index) * 2 + getEta(index) + 1;
-    case 2:
-      return ((getPhi(index) + 1) % nphi_) * 2 + getEta(index);
-    case 3:
-      return ((getPhi(index) + 1) % nphi_) * 2 + getEta(index) + 1;
-    default:
-      dbgCout() << "Impossible object count!" << std::endl;
-      exit(0);
-  }
-}
-
-template <typename T>
-int l1ct::tdr_regionizer::Pipe<T>::getPipeIndexForObject(unsigned int index) {
-  switch (getCount(index)) {
-    case 0:
-      return getPhi(index);
-    case 1:
-      if (getPhiOverlap(index) && !getEtaOverlap(index))
-        return (getPhi(index) + 1) % nphi_;
-      else
-        return getPhi(index);
-    case 2:
-    case 3:
-      return (getPhi(index) + 1) % nphi_;
-    default:
-      dbgCout() << "Impossible object count!" << std::endl;
-      exit(0);
-  }
-}
-
-template <typename T>
-l1ct::tdr_regionizer::Regionizer<T>::Regionizer(
-    unsigned int neta, unsigned int nregions, unsigned int maxobjects, int etaoffset, int etawidth, int nclocks)
+inline unsigned int l1ct::tdr_regionizer::Buffer<l1ct::TkObjEmu>::nextObjClk(unsigned int ndup) {
+  assert(ndup == 1);
```

Comment:
I'm a little confused by this. You take `ndup` as a parameter... but then assert it has to be 1 or crash? Why is this done this way?

Reply:
Basically it's that ndup != 1 is not implemented yet. In particular, it hasn't been implemented in the corresponding firmware, and there will likely never be a need for it. This function is the template specialization for TkObjEmu, which is a bit special since two objects are sent in three transfers, unlike the others which are one object per transfer.

Reply:
Sure. I'll update it.

Reply:
Cherry-picked here the commit that throws the exception.

For the reviewers: it's a std::exception and not a cms::Exception since this emulator code is also compiled standalone outside CMSSW.